[emit] Implement emit functions for bitfield move and extract ops

dongju-chae opened the pull request:
This PR implements functions for bitfield move and extract ops.
- sbfm, bfm, ubfm, extr

Note that These ops are commonly used as aliases
- lsl (imm), lsr (imm), asr (imm), ror (imm), ...

Signed-off-by: Dongju Chae <dongju.chae@samsung.com>

myungjoo commented on orc/orcarm.c:
```diff
@@ -1016,3 +1016,201 @@ orc_arm64_emit_am (OrcCompiler *p, OrcArm64RegBits bits, OrcArm64DP opcode,
+
+#define arm64_code_bfm(b,opcode,immr,imms,Rn,Rd) (0x19000000 | (((b==64)&0x1)<<31) | \
+    (((opcode)&0x3)<<29) | (((b==64)&0x1)<<22) | (((immr)&0x3f)<<16) | (((imms)&0x3f)<<10) | \
+    (((Rn)&0x1f)<<5) | ((Rd)&0x1f))
```
Protect b by replacing ```(b==64)``` with ```((b)==64)```

dongju-chae replied:
Got it.

myungjoo replied:
You still have ```(((b==64)&0x1)<<22)```

dongju-chae replied:
Oh, sorry

myungjoo commented on orc/orcarm.c:
```diff
@@ -1016,3 +1016,201 @@ orc_arm64_emit_am (OrcCompiler *p, OrcArm64RegBits bits, OrcArm64DP opcode,
+      break;
+    default:
+      ORC_COMPILER_ERROR(p, "unknown opcode %d", opcode);
+      return;
```
Our secure-coding checker will complain the usage of sprintf. Consider using snprintf with the static length (63) for both opt_immr and opt_imms.

dongju-chae replied:
Got it

myungjoo commented on orc/orcarm.c:
```diff
@@ -1016,3 +1016,201 @@ orc_arm64_emit_am (OrcCompiler *p, OrcArm64RegBits bits, OrcArm64DP opcode,
+ */
+
+#define arm64_code_extr(b,Rm,imms,Rn,Rd) (0x13800000 | (((b==64)&0x1)<<31) | \
+    (((b==64)&0x1)<<22) | (((Rm)&0x1f)<<16) | (((imms)&0x3f)<<10) | \
```
wrapping b with () is also recommended here.

dongju-chae replied:
Updated

myungjoo commented:
Sorry. I thought the size of snprintf does not count the terminating null byte; however, the manual says:
```
       The functions snprintf() and vsnprintf() write at most size bytes (including the terminating null byte ('\0')) to str.
```

dongju-chae commented:
Then, `-1` can be removed.

Commit:
[emit] Implement emit functions for bitfield move ops

This commit implements functions for bitfield move ops.
- sbfm, bfm, ubfm

Note that These ops are commonly used as aliases
- lsl (imm), lsr (imm), asr (imm), ...

Signed-off-by: Dongju Chae <dongju.chae@samsung.com>

Commit:
[emit] Implement emit functions for extract op

This commit implements an emit function for extract op, extr.
Note that this op is usually used as the alias, ror (imm).

Signed-off-by: Dongju Chae <dongju.chae@samsung.com>

myungjoo reviewed (approved):
LGTM
